src/core_test/mpas_test_core_halo_exch.F

```diff
@@ -1370,5 +1383,227 @@ subroutine test_core_halo_exch_validate_fields(domain, threadErrs, err)!{{{
+   !>  Cells marked with a 2 should get updated via halo layer 1.
+   !>  Cells marked with a 3 should get updated via halo layer 2, etc.
+   !>  
+   !-----------------------------------------------------------------------
+   function findExteriorCells(nCellsSolve, nCells, cellsOnCell, edgesOnCell, nHaloLayers) &
+      result(exteriorCells)
+      integer, intent(in) :: nCellsSolve !< the number of cells in this block
+      integer, intent(in) :: nCells !< total number of cells (cells in this block plus halo cells)
+      integer, dimension(:,:), intent(in) :: cellsOnCell !< array with adjacent cells for each cell
+      integer, dimension(:), intent(in) :: edgesOnCell !< array with edges for each cell
+      integer, intent(in) :: nHaloLayers  !< the number of halo layers
+
+      integer, dimension(:), allocatable :: exteriorCells
+      integer nInterior, nEdge, nLayers
+
+      allocate(exteriorCells(nCells))
+      exteriorCells(1:nCellsSolve) = 0 !< mark all owned cells as interior
+      exteriorCells(nCellsSolve+1:nCells) = 1 !< mark all halo cells as edge
+      nInterior = 0
+      nEdge = 0
+#ifdef HALO_EXCH_DEBUG_VERBOSE
+      call mpas_log_write(' halo cellsOnCell($i x $i)', &
+         intArgs=(/size(cellsOnCell, dim=1), size(cellsOnCell, dim=2)/))
+#endif
+      
+      ! At this point, only halo cells are marked 1, and all owned cells are marked 0
+      ! for each halo layer, mark cells adjacent to already marked cells with next highest marker
+      do nLayers = 1, nHaloLayers
+         nEdge = nEdge + mark_interior_cells(exteriorCells(1:nCells), nLayers, cellsOnCell, edgesOnCell)
+      end do
+
+      ! Now, interior cells that are adjacent to halo cells are marked 2 (edge cells)
+      ! Mark all cells adjacent to edge cells as 3
+      !nEdge = nEdge + mark_interior_cells(exteriorCells(1:nCells), 2, cellsOnCell, edgesOnCell)
+
+      ! Now, interior cells that are adjacent to edge cells are marked 3
+      ! Mark cells adjacent to cells which are adjecent to edge cells as 4
+      !nEdge = nEdge + mark_interior_cells(exteriorCells(1:nCells), 3, cellsOnCell, edgesOnCell)
+      nInterior = nCellsSolve - nEdge
+
+#ifdef HALO_EXCH_DEBUG_VERBOSE
+      call mpas_log_write(' halo nInterior:$i nEdge:$i', intArgs=(/nInterior, nEdge/))
+#endif
+   end function findExteriorCells
+
+   !***********************************************************************
+   !> \brief   MPAS Test Core halo adjoint exchange
+   !> \details 
+   !>  This routine applies the halo adjoint function to a 2D array and verifies
+   !>  interior cell's values don't change, and border cell's values are modified.
+   !>  It assumes a halo exchange has already been applied.
+   !>  Returns 0 on success, non-0 on failure.
+   !-----------------------------------------------------------------------
+   subroutine test_halo_adj_exch_fields(domain, threadErrs, err)
+      type (domain_type), intent(inout) :: domain
```

Comment:
Here and elsewhere, can you add a blank line after the `subroutine` statement before the argument declarations?

Reply:
Yes.

Reply:
I think the `check_adjoint_values` function also needs a blank line.